packages/@aws-cdk/aws-route53/lib/cross-account-zone-delegation-handler/index.ts

```diff
@@ -0,0 +1,76 @@
+// eslint-disable-next-line import/no-extraneous-dependencies
+import { Credentials, Route53, STS } from 'aws-sdk';
+
+interface ResourceProperties {
+  AssumeRoleArn: string,
+  ParentZoneName: string,
+  RecordName: string,
+  NameServers: string[],
+  TTL: number,
+}
+
+export async function handler(event: AWSLambda.CloudFormationCustomResourceEvent) {
+  const resourceProps = event.ResourceProperties as unknown as ResourceProperties;
+
+  switch (event.RequestType) {
+    case 'Create':
+    case 'Update':
+      return cfnEventHandler(resourceProps, false);
+    case 'Delete':
+      return cfnEventHandler(resourceProps, true);
+  }
+}
+
+async function cfnEventHandler(props: ResourceProperties, isDeleteEvent: boolean) {
+  const { AssumeRoleArn, ParentZoneName, RecordName, NameServers, TTL } = props;
+
+  const credentials = await getCrossAccountCredentials(AssumeRoleArn);
+  const route53 = new Route53({ credentials });
+
+  const { HostedZones } = await route53.listHostedZonesByName({ DNSName: ParentZoneName }).promise();
+  if (HostedZones.length < 1) {
```

Comment:
As per [these docs](https://docs.aws.amazon.com/Route53/latest/APIReference/API_ListHostedZonesByName.html) and my understanding, the `listHostedZonesByName` API works similarly to `listHostedZones` API with a difference of sorting. And for the `DNSName` prop:
>For your first request to ListHostedZonesByName, include the dnsname parameter only if you want to specify the name of the first hosted zone in the response.

Considering that, I decided to go ahead with this approach of using name api instead of just normal list api. But when trying more things around it here is the result of some api calls for a account containing these zones:
1. myzone.com
2. sub.myzone.com
3. anotherzone.com
4. otherprivatezone.com => private

```
$ aws route53 list-hosted-zones-by-name --dns-name myzone.com
{
    "HostedZones": [
        {
            "Id": "/hostedzone/Z086460468SYA64PZBV0",
            "Name": "myzone.com.",
            "CallerReference": "7bfc3a5d-201c-43f9-b142-69a670c41e70",
            "Config": {
                "Comment": "",
                "PrivateZone": false
            },
            "ResourceRecordSetCount": 2
        },
        {
            "Id": "/hostedzone/Z08499033HVQ5LQW1NR59",
            "Name": "sub.myzone.com.",
            "CallerReference": "7e954789-39a4-4120-9ad7-826cf1a8a0fa",
            "Config": {
                "Comment": "",
                "PrivateZone": false
            },
            "ResourceRecordSetCount": 2
        },
        {
            "Id": "/hostedzone/Z05723083AGKK4YNDVF4G",
            "Name": "otherprivatezone.com.",
            "CallerReference": "41e32ed0-73c3-4566-8e58-324d95dbe630",
            "Config": {
                "Comment": "",
                "PrivateZone": true
            },
            "ResourceRecordSetCount": 2
        }
    ],
    "DNSName": "myzone.com",
    "IsTruncated": false,
    "MaxItems": "100"
}
$ aws route53 list-hosted-zones-by-name --dns-name sdfsdfe.com
{
    "HostedZones": [],
    "DNSName": "sdfsdfe.com",
    "IsTruncated": false,
    "MaxItems": "100"
}
$ aws route53 list-hosted-zones-by-name --dns-name zone.com
{
    "HostedZones": [],
    "DNSName": "zone.com",
    "IsTruncated": false,
    "MaxItems": "100"
}
$ aws route53 list-hosted-zones-by-name --dns-name my.com
{
    "HostedZones": [
        {
            "Id": "/hostedzone/Z086460468SYA64PZBV0",
            "Name": "myzone.com.",
            "CallerReference": "7bfc3a5d-201c-43f9-b142-69a670c41e70",
            "Config": {
                "Comment": "",
                "PrivateZone": false
            },
            "ResourceRecordSetCount": 2
        },
        {
            "Id": "/hostedzone/Z08499033HVQ5LQW1NR59",
            "Name": "sub.myzone.com.",
            "CallerReference": "7e954789-39a4-4120-9ad7-826cf1a8a0fa",
            "Config": {
                "Comment": "",
                "PrivateZone": false
            },
            "ResourceRecordSetCount": 2
        },
        {
            "Id": "/hostedzone/Z05723083AGKK4YNDVF4G",
            "Name": "otherprivatezone.com.",
            "CallerReference": "41e32ed0-73c3-4566-8e58-324d95dbe630",
            "Config": {
                "Comment": "",
                "PrivateZone": true
            },
            "ResourceRecordSetCount": 2
        }
    ],
    "DNSName": "my.com",
    "IsTruncated": false,
    "MaxItems": "100"
}
 
```

Do you think we should just use the `listHostedZone` api instead (which is paginated) and filter our hosted zone from the result? Also, one thing which got missed is we are not checking if the returned zone is public or not. We should do that too.

Reply:
Added usage of `delegationZone` and `parentHostedZoneId`.